### PR TITLE
feat(triple): support generic call for Triple protocol

### PR DESCRIPTION
**Pull request by Tsukikage7**
### Description

This PR implements **Triple protocol generic call** support for dubbo-go, enabling Go clients
 to invoke Java Dubbo services without pre-generated stubs and vice versa.

For detailed design documentation, please refer to: [Dubbo-go Triple Generic
Call](https://www.tsukikage7.com/blog/dubbo-go-triple-generic-call/)

**Related PR:** apache/dubbo-go-samples#1016 provides usage examples for Triple protocol
generic call.

#### Background

Generic call (泛化调用) is a core feature in Dubbo ecosystem that allows service invocation
without IDL/interface definitions. While dubbo-go already supports generic call for the Dubbo
 protocol, Triple protocol generic call was missing, making it impossible for Go and Java
services to interoperate via Triple protocol in non-IDL mode.

#### Design Overview

The key challenge is that Java Dubbo Triple protocol uses a specific wire format for non-IDL
mode:

```
┌─────────────────────────────────────────────────────────────┐
│                    Wire Format                              │
├─────────────────────────────────────────────────────────────┤
│  Content-Type: application/proto                            │
│                                                             │
│  Request:  TripleRequestWrapper (protobuf)                  │
│            ├── serializeType: "hessian2"                    │
│            ├── args: [hessian2(arg1), hessian2(arg2), ...]  │
│            └── argTypes: ["java.lang.String", ...]          │
│                                                             │
│  Response: TripleResponseWrapper (protobuf)                 │
│            ├── serializeType: "hessian2"                    │
│            ├── data: hessian2(result)                       │
│            └── type: "org.apache.dubbo.samples.User"        │
└─────────────────────────────────────────────────────────────┘
```

This is a **two-layer serialization**: protobuf wrapper on the outside, hessian2 data on the
inside.

#### Implementation Details

**1. New `WrapperCodec` Interface (`codec.go`)**

```go
type WrapperCodec interface {
    Codec
    WireCodecName() string  // Returns "proto" for Content-Type
}
```

**2. New `protoWrapperCodec` (`codec.go`)**

Client-side codec that wraps request arguments in `TripleRequestWrapper` and unwraps response
 from `TripleResponseWrapper`.

**3. Enhanced `protoBinaryCodec` (`codec.go`)**

Server-side enhancement to handle incoming wrapped requests.

**4. Generic Filter Enhancement (`filter/generic/filter.go`)**

Added `parameterRawValues` support for Triple protocol in both `isCallingToGenericService`
and `isMakingAGenericCall` branches.

**5. Server-side Generic Call Handler (`protocol/triple/server.go`)**

Auto-registers `$invoke` method for Triple protocol services via `buildGenericMethodInfo()`.

#### Testing

- **Unit Tests**: 631 lines of comprehensive tests covering `protoWrapperCodec`,
`protoBinaryCodec`, various data types, and edge cases.
- **Integration Tests**: All scenarios passed (Go↔Java, Go↔Go via Triple protocol).

#### Compatibility

- **Backward Compatible**: Existing IDL-based Triple calls are not affected
- **Cross-Language**: Fully compatible with Java Dubbo 3.x Triple protocol
- **Dubbo Protocol**: Existing Dubbo protocol generic call still works as before

### Checklist
- [x] I confirm the target branch is `main`
- [x] Code has passed local testing
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Tsukikage7** [comment]
> **📌 Related Sample PR**
> 
> apache/dubbo-go-samples#1016 demonstrates the usage of `NewGenericService` API added in this PR.
> 
> The sample PR's CI will pass after this PR is merged to main.

**Review by No-SilverBullet — APPROVED** [review]
LGTM :)

**Comment by Tsukikage7** [comment]
@marsevilspirit @Alanxtl Need Review pls.

File: protocol/triple/triple_protocol/codec.go [hunk 1]
```diff
@@ -114,11 +118,62 @@ func (c *protoBinaryCodec) Marshal(message any) ([]byte, error) {
...
 }
 
+// unmarshalWrappedMessage handles both TripleResponseWrapper and TripleRequestWrapper formats.
+// It determines the format by checking if message is a slice (request) or not (response).
+func (c *protoBinaryCodec) unmarshalWrappedMessage(data []byte, message any) error {
+	hessianCodec := &hessian2Codec{}
+
+	// Check if message is a slice - if so, it's a request with multiple args
+	params, isSlice := message.([]any)
+	if isSlice {
```
Comment (Alanxtl) [inline protocol/triple/triple_protocol/codec.go]:
```suggestion
	if params, isSlice := message.([]any); isSlice {
```

File: protocol/triple/triple_protocol/codec.go [hunk 2]
```diff
@@ -227,29 +322,51 @@ func (c *protoWrapperCodec) Marshal(message any) ([]byte, error) {
...
 func (c *protoWrapperCodec) Unmarshal(binary []byte, message any) error {
-	var params []any
-	var ok bool
-	params, ok = message.([]any)
-	if !ok {
-		params = []any{message}
-	}
+	// Check if message is a slice - if so, it's a request with multiple args
+	params, isSlice := message.([]any)
+	if isSlice {
```
Comment (Alanxtl) [inline protocol/triple/triple_protocol/codec.go]: ditto

File: protocol/triple/triple.go [hunk 3]
```diff
@@ -141,6 +149,18 @@ func (tp *TripleProtocol) Destroy() {
...
+func isGenericCall(generic string) bool {
+	if generic == "" {
+		return false
+	}
+	lowerGeneric := strings.ToLower(generic)
+	return lowerGeneric == constant.GenericSerializationDefault ||
+		lowerGeneric == constant.GenericSerializationGson ||
+		lowerGeneric == constant.GenericSerializationProtobuf ||
+		lowerGeneric == constant.GenericSerializationProtobufJson
+}
```
Comment (Alanxtl) [inline protocol/triple/triple.go]: use `strings.EqualFold` instead, no need to `toLower` and `==`

File: protocol/triple/server.go [hunk 4]
```diff
@@ -503,51 +503,101 @@ func createServiceInfoWithReflection(svc common.RPCService) *common.ServiceInfo
...
 		}
-		methodInfos = append(methodInfos, methodInfo)
 	}
 
-	// only support no-idl mod call unary
-	genericMethodInfo := common.MethodInfo{
-		Name: "$invoke",
-		Type: constant.CallUnary,
+	// Add $invoke method for generic call support
+	methodInfos = append(methodInfos, buildGenericMethodInfo())
```
Comment (Alanxtl) [inline protocol/triple/server.go:513]: this `methodInfos` is appended inside the loop in line 508 and appended outside the loop in line 513, is it correct?

File: server/server.go [hunk 5]
```diff
@@ -176,6 +179,61 @@ func (s *Server) genSvcOpts(handler any, info *common.ServiceInfo, opts ...Servi
...
+		return true
+	default:
+		return false
+	}
+}
+
+// isReflectNil safely checks if a reflect.Value is nil.
+func isReflectNil(v reflect.Value) bool {
+	return isNillable(v) && v.IsNil()
+}
```
Comment (Alanxtl) [inline server/server.go:195]: is this method redefined in the protocol/triple/server.go:594?

File: server/server.go [hunk 6]
```diff
@@ -176,6 +179,61 @@ func (s *Server) genSvcOpts(handler any, info *common.ServiceInfo, opts ...Servi
...
+// createReflectionMethodFunc creates a MethodFunc that calls the given method via reflection.
+func createReflectionMethodFunc(method reflect.Method) func(ctx context.Context, args []any, handler any) (any, error) {
+	return func(ctx context.Context, args []any, handler any) (any, error) {
+		return CallMethodByReflection(ctx, method, handler, args)
+	}
+}
+
 // Add a method with a name of a different first-letter case
 // to achieve interoperability with java
 // TODO: The method name case sensitivity in Dubbo-java should be addressed.
```
Comment (Alanxtl) [inline server/server.go:239]: I'm wondering is this todo solved?

File: protocol/triple/server.go (hunk 4 above)
Comment (Tsukikage7) [reply protocol/triple/server.go:513]: Yeah, that's on purpose. Inside the loop we're adding all the regular service methods, then after the loop we add $invoke once for generic call support. Makes sure we don't accidentally register it multiple times.

File: server/server.go (hunk 5 above)
Comment (Tsukikage7) [reply server/server.go:195]: Good catch. They do look similar but live in different packages:

server/server.go has CallMethodByReflection (exported, shared)
protocol/triple/server.go has isReflectValueNil (local helper)
Thought about extracting a common util, but protocol/triple importing server package might cause circular deps. It's just a few lines so I think we're fine for now. Happy to refactor if you think it's worth it.

File: server/server.go (hunk 6 above)
Comment (Tsukikage7) [reply server/server.go:239]: Partially. The enhanceServiceInfo function now registers both GetUser and getUser style methods, so Go and Java can find each other. But the real fix should be on dubbo-java side to handle case properly. Kept the TODO as a reminder - maybe we can push for a fix upstream later.

File: protocol/triple/server.go [hunk 7]
```diff
@@ -503,51 +503,101 @@ func createServiceInfoWithReflection(svc common.RPCService) *common.ServiceInfo
...
+				&[]hessian.Object{},                      // args *[]hessian.Object
+			}
+		},
+	}
+}
 
-	return &info
+// isReflectValueNil safely checks if a reflect.Value is nil.
+// It first checks if the value's kind supports nil checking to avoid panic.
+func isReflectValueNil(v reflect.Value) bool {
```
Comment (AlexStocks) [inline protocol/triple/server.go:594]: 在 case 1 中增加对 UnsafePointer 的检查：
func isReflectValueNil(v reflect.Value) bool {
    switch v.Kind() {
    case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, 
         reflect.Pointer, reflect.Slice, reflect.UnsafePointer:
        return v.IsNil()
    default:
        return false
    }
}

**Comment by AlexStocks** [comment]
当前代码我留了一个 comment，请改进。另外，请给其中一些函数添加 unit test。

File: protocol/triple/server.go (hunk 7 above)
Comment (Tsukikage7) [reply protocol/triple/server.go:594]: done

File: server/server.go (hunk 6 above)
Comment (AlexStocks) [reply server/server.go:239]: 建个issue，记下这个 todo 吧

**Review by Alanxtl — APPROVED** [review]
